diff -uraN linux-2.6.0-test7/drivers/char/Kconfig linux-2.6.0-test7-fusion/drivers/char/Kconfig
--- linux-2.6.0-test7/drivers/char/Kconfig	2003-10-08 21:24:01.000000000 +0200
+++ linux-2.6.0-test7-fusion/drivers/char/Kconfig	2003-10-14 09:44:07.000000000 +0200
@@ -971,5 +971,10 @@
 	  out to lunch past a certain margin.  It can reboot the system
 	  or merely print a warning.
 
-endmenu
+config FUSION_DEVICE
+	tristate "Fusion Kernel Device (for DirectFB)"
+	help
+	  The Fusion Kernel Device is needed to use DirectFB's multi
+	  application core.
 
+endmenu
diff -uraN linux-2.6.0-test7/drivers/char/Makefile linux-2.6.0-test7-fusion/drivers/char/Makefile
--- linux-2.6.0-test7/drivers/char/Makefile	2003-10-08 21:24:50.000000000 +0200
+++ linux-2.6.0-test7-fusion/drivers/char/Makefile	2003-10-14 09:44:07.000000000 +0200
@@ -78,6 +78,8 @@
 obj-$(CONFIG_PCMCIA) += pcmcia/
 obj-$(CONFIG_IPMI_HANDLER) += ipmi/
 
+obj-$(CONFIG_FUSION_DEVICE) += fusion/
+
 obj-$(CONFIG_HANGCHECK_TIMER) += hangcheck-timer.o
 
 # Files generated that shall be removed upon make clean
