diff -Naur linux-2.4.20-rc1/drivers/char/Config.in linux-2.4.20-rc1-fusion/drivers/char/Config.in
--- linux-2.4.20-rc1/drivers/char/Config.in	2002-11-01 13:32:31.000000000 +0100
+++ linux-2.4.20-rc1-fusion/drivers/char/Config.in	2002-11-02 03:11:39.000000000 +0100
@@ -228,6 +228,7 @@
    dep_tristate 'Intel i8x0 Random Number Generator support' CONFIG_INTEL_RNG $CONFIG_PCI
 fi
 tristate '/dev/nvram support' CONFIG_NVRAM
+tristate 'Fusion Kernel Device' CONFIG_FUSION_DEVICE
 tristate 'Enhanced Real Time Clock Support' CONFIG_RTC
 if [ "$CONFIG_IA64" = "y" ]; then
    bool 'EFI Real Time Clock Services' CONFIG_EFI_RTC
diff -Naur linux-2.4.20-rc1/drivers/char/Makefile linux-2.4.20-rc1-fusion/drivers/char/Makefile
--- linux-2.4.20-rc1/drivers/char/Makefile	2002-11-01 13:32:31.000000000 +0100
+++ linux-2.4.20-rc1-fusion/drivers/char/Makefile	2002-11-02 04:18:08.000000000 +0100
@@ -268,6 +268,12 @@
   obj-y += mwave/mwave.o
 endif
 
+subdir-$(CONFIG_FUSION_DEVICE) += fusion
+ifeq ($(CONFIG_FUSION_DEVICE),y)
+  obj-y += fusion/fusion.o
+endif
+
+
 include $(TOPDIR)/Rules.make
 
 fastdep:
diff -Naur linux-2.4.20-rc1/include/linux/miscdevice.h linux-2.4.20-rc1-fusion/include/linux/miscdevice.h
--- linux-2.4.20-rc1/include/linux/miscdevice.h	2002-11-01 13:33:02.000000000 +0100
+++ linux-2.4.20-rc1-fusion/include/linux/miscdevice.h	2002-11-02 03:51:39.000000000 +0100
@@ -14,6 +14,7 @@
 #define PC110PAD_MINOR 9
 #define ADB_MOUSE_MINOR 10
 #define MK712_MINOR 15			/* MK712 touch screen */
+#define FUSION_MINOR		 23	/* Fusion Device      */
 #define WATCHDOG_MINOR		130	/* Watchdog timer     */
 #define TEMP_MINOR		131	/* Temperature Sensor */
 #define RTC_MINOR 135
