diff -Naur linux-2.4.20-pre11/drivers/char/Config.in linux-2.4.20-pre11-fusion/drivers/char/Config.in
--- linux-2.4.20-pre11/drivers/char/Config.in	2002-10-16 14:04:12.000000000 +0200
+++ linux-2.4.20-pre11-fusion/drivers/char/Config.in	2002-10-28 18:11:54.000000000 +0100
@@ -244,6 +244,7 @@
    dep_tristate 'Intel i8x0 Random Number Generator support' CONFIG_INTEL_RNG $CONFIG_PCI
 fi
 tristate '/dev/nvram support' CONFIG_NVRAM
+tristate 'Fusion Kernel Device' CONFIG_FUSION_DEVICE
 tristate 'Enhanced Real Time Clock Support' CONFIG_RTC
 if [ "$CONFIG_IA64" = "y" ]; then
    bool 'EFI Real Time Clock Services' CONFIG_EFI_RTC
diff -Naur linux-2.4.20-pre11/drivers/char/Makefile linux-2.4.20-pre11-fusion/drivers/char/Makefile
--- linux-2.4.20-pre11/drivers/char/Makefile	2002-10-16 14:04:12.000000000 +0200
+++ linux-2.4.20-pre11-fusion/drivers/char/Makefile	2002-10-28 18:11:54.000000000 +0100
@@ -288,6 +288,9 @@
   obj-y += mwave/mwave.o
 endif
 
+subdir-$(CONFIG_FUSION_DEVICE) += fusion
+
+
 include $(TOPDIR)/Rules.make
 
 fastdep:
diff -Naur linux-2.4.20-pre11/drivers/char/misc.c linux-2.4.20-pre11-fusion/drivers/char/misc.c
--- linux-2.4.20-pre11/drivers/char/misc.c	2002-10-16 14:02:02.000000000 +0200
+++ linux-2.4.20-pre11-fusion/drivers/char/misc.c	2002-10-28 18:10:48.000000000 +0100
@@ -77,6 +77,7 @@
 extern int tosh_init(void);
 extern int i8k_init(void);
 extern int lcd_init(void);
+extern int fusion_init(void);
 
 static int misc_read_proc(char *buf, char **start, off_t offset,
 			  int len, int *eof, void *private)
@@ -285,6 +286,9 @@
 #ifdef CONFIG_I8K
 	i8k_init();
 #endif
+#ifdef CONFIG_FUSION_DEVICE
+	fusion_init();
+#endif
 	if (devfs_register_chrdev(MISC_MAJOR,"misc",&misc_fops)) {
 		printk("unable to get major %d for misc devices\n",
 		       MISC_MAJOR);
diff -Naur linux-2.4.20-pre11/include/linux/miscdevice.h linux-2.4.20-pre11-fusion/include/linux/miscdevice.h
--- linux-2.4.20-pre11/include/linux/miscdevice.h	2002-10-16 14:03:45.000000000 +0200
+++ linux-2.4.20-pre11-fusion/include/linux/miscdevice.h	2002-10-28 18:11:54.000000000 +0100
@@ -14,6 +14,7 @@
 #define PC110PAD_MINOR 9
 #define ADB_MOUSE_MINOR 10
 #define MK712_MINOR 15			/* MK712 touch screen */
+#define FUSION_MINOR		 23	/* Fusion Device      */
 #define WATCHDOG_MINOR		130	/* Watchdog timer     */
 #define TEMP_MINOR		131	/* Temperature Sensor */
 #define RTC_MINOR 135
