diff -uraN linux-2.4.20/drivers/char/Config.in linux-2.4.20-fusion/drivers/char/Config.in
--- linux-2.4.20/drivers/char/Config.in	2002-11-29 13:18:31.000000000 +0100
+++ linux-2.4.20-fusion/drivers/char/Config.in	2002-12-05 18:39:08.000000000 +0100
@@ -248,6 +248,7 @@
 fi
 dep_tristate 'AMD 76x native power management (Experimental)' CONFIG_AMD_PM768 $CONFIG_PCI
 tristate '/dev/nvram support' CONFIG_NVRAM
+tristate 'Fusion Kernel Device' CONFIG_FUSION_DEVICE
 tristate 'Enhanced Real Time Clock Support' CONFIG_RTC
 if [ "$CONFIG_IA64" = "y" ]; then
    bool 'EFI Real Time Clock Services' CONFIG_EFI_RTC
diff -uraN linux-2.4.20/drivers/char/Makefile linux-2.4.20-fusion/drivers/char/Makefile
--- linux-2.4.20/drivers/char/Makefile	2002-11-29 13:18:31.000000000 +0100
+++ linux-2.4.20-fusion/drivers/char/Makefile	2002-12-05 18:39:08.000000000 +0100
@@ -289,6 +289,12 @@
   obj-y += mwave/mwave.o
 endif
 
+subdir-$(CONFIG_FUSION_DEVICE) += fusion
+ifeq ($(CONFIG_FUSION_DEVICE),y)
+  obj-y += fusion/fusion.o
+endif
+
+
 include $(TOPDIR)/Rules.make
 
 fastdep:
