diff -uraN linux-2.4.22/drivers/char/Config.in linux-2.4.22-fusion/drivers/char/Config.in
--- linux-2.4.22/drivers/char/Config.in	2003-08-25 13:44:41.000000000 +0200
+++ linux-2.4.22-fusion/drivers/char/Config.in	2003-10-14 09:21:04.000000000 +0200
@@ -269,6 +269,7 @@
 fi
 dep_tristate 'AMD 76x native power management (Experimental)' CONFIG_AMD_PM768 $CONFIG_PCI
 tristate '/dev/nvram support' CONFIG_NVRAM
+tristate 'Fusion Kernel Device' CONFIG_FUSION_DEVICE
 tristate 'Enhanced Real Time Clock Support' CONFIG_RTC
 if [ "$CONFIG_IA64" = "y" ]; then
    bool 'EFI Real Time Clock Services' CONFIG_EFI_RTC
diff -uraN linux-2.4.22/drivers/char/Makefile linux-2.4.22-fusion/drivers/char/Makefile
--- linux-2.4.22/drivers/char/Makefile	2003-08-25 13:44:41.000000000 +0200
+++ linux-2.4.22-fusion/drivers/char/Makefile	2003-10-14 09:21:05.000000000 +0200
@@ -313,6 +313,12 @@
   obj-y += ipmi/ipmi.o
 endif
 
+subdir-$(CONFIG_FUSION_DEVICE) += fusion
+ifeq ($(CONFIG_FUSION_DEVICE),y)
+  obj-y += fusion/fusion.o
+endif
+
+
 include $(TOPDIR)/Rules.make
 
 fastdep:
