diff -uraN linux-2.6.0-test1/drivers/char/Kconfig linux-2.6.0-test1-fusion/drivers/char/Kconfig
--- linux-2.6.0-test1/drivers/char/Kconfig	2003-07-23 23:24:11.000000000 +0200
+++ linux-2.6.0-test1-fusion/drivers/char/Kconfig	2003-07-24 02:37:03.000000000 +0200
@@ -1006,5 +1006,10 @@
 	  out to lunch past a certain margin.  It can reboot the system
 	  or merely print a warning.
 
-endmenu
+config FUSION_DEVICE
+	tristate "Fusion Kernel Device (for DirectFB)"
+	help
+	  The Fusion Kernel Device is needed to use DirectFB's multi
+	  application core.
 
+endmenu
diff -uraN linux-2.6.0-test1/drivers/char/Makefile linux-2.6.0-test1-fusion/drivers/char/Makefile
--- linux-2.6.0-test1/drivers/char/Makefile	2003-07-23 23:29:37.000000000 +0200
+++ linux-2.6.0-test1-fusion/drivers/char/Makefile	2003-07-24 02:37:03.000000000 +0200
@@ -78,6 +78,8 @@
 obj-$(CONFIG_PCMCIA) += pcmcia/
 obj-$(CONFIG_IPMI_HANDLER) += ipmi/
 
+obj-$(CONFIG_FUSION_DEVICE) += fusion/
+
 obj-$(CONFIG_HANGCHECK_TIMER) += hangcheck-timer.o
 
 # Files generated that shall be removed upon make clean
diff -uraN linux-2.6.0-test1/drivers/char/fusion/Makefile linux-2.6.0-test1-fusion/drivers/char/fusion/Makefile
--- linux-2.6.0-test1/drivers/char/fusion/Makefile	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.0-test1-fusion/drivers/char/fusion/Makefile	2003-07-24 02:37:03.000000000 +0200
@@ -0,0 +1,3 @@
+obj-$(CONFIG_FUSION_DEVICE) := fusion.o
+
+fusion-objs := call.o fifo.o fusiondev.o fusionee.o list.o property.o reactor.o ref.o skirmish.o
