diff -uraN linux-2.5.66/drivers/char/Kconfig linux-2.5.66-fusion/drivers/char/Kconfig
--- linux-2.5.66/drivers/char/Kconfig	2003-04-05 03:58:16.000000000 +0200
+++ linux-2.5.66-fusion/drivers/char/Kconfig	2003-04-05 04:35:22.000000000 +0200
@@ -980,5 +980,10 @@
 	  out to lunch past a certain margin.  It can reboot the system
 	  or merely print a warning.
 
-endmenu
+config FUSION_DEVICE
+	tristate "Fusion Kernel Device (for DirectFB)"
+	help
+	  The Fusion Kernel Device is needed to use DirectFB's multi
+	  application core.
 
+endmenu
diff -uraN linux-2.5.66/drivers/char/Makefile linux-2.5.66-fusion/drivers/char/Makefile
--- linux-2.5.66/drivers/char/Makefile	2003-04-05 03:59:12.000000000 +0200
+++ linux-2.5.66-fusion/drivers/char/Makefile	2003-04-05 04:35:22.000000000 +0200
@@ -77,6 +77,8 @@
 obj-$(CONFIG_PCMCIA) += pcmcia/
 obj-$(CONFIG_IPMI_HANDLER) += ipmi/
 
+obj-$(CONFIG_FUSION_DEVICE) += fusion/
+
 obj-$(CONFIG_HANGCHECK_TIMER) += hangcheck-timer.o
 
 # Files generated that shall be removed upon make clean
diff -uraN linux-2.5.66/include/linux/miscdevice.h linux-2.5.66-fusion/include/linux/miscdevice.h
--- linux-2.5.66/include/linux/miscdevice.h	2003-01-17 03:22:18.000000000 +0100
+++ linux-2.5.66-fusion/include/linux/miscdevice.h	2003-04-05 04:35:22.000000000 +0200
@@ -13,6 +13,7 @@
 #define APOLLO_MOUSE_MINOR 7
 #define PC110PAD_MINOR 9
 /*#define ADB_MOUSE_MINOR 10	FIXME OBSOLETE */
+#define FUSION_MINOR		23	/* Fusion Device */
 #define WATCHDOG_MINOR		130	/* Watchdog timer     */
 #define TEMP_MINOR		131	/* Temperature Sensor */
 #define RTC_MINOR 135
